### PR TITLE
logs storage service and RPC definitions

### PR DESCRIPTION
**Pull request by sayan-biswas**
This PR separates the service definition of logs from the results in the Proto file. Also the logs service registration is made optional at the API server and the watcher can negotiate the availability of logs service through a reflection client and enable/disable sending logs.

### CODE REVIEW

File: cmd/watcher/main.go [hunk 1]
```diff
@@ -58,6 +59,7 @@ var (
 	authToken               = flag.String("token", "", "Authentication token to use in requests. If not specified, on-cluster configuration is assumed.")
 	completedRunGracePeriod = flag.Duration("completed_run_grace_period", 0, "Grace period duration before Runs should be deleted. If 0, Runs will not be deleted. If < 0, Runs will be deleted immediately.")
 	threadiness             = flag.Int("threadiness", controller.DefaultThreadsPerController, "Number of threads (Go routines) allocated to each controller")
+	logsAPI                 = flag.Bool("logs_api", true, "Disable sending logs. If not set, the logs will be sent only if server support API for it")
```
Comment (adambkaplan) [inline cmd/watcher/main.go:62]: We are likely going to need to revisit this logic in the future. I like the fact that watcher will only forward logs if the API server publishes the service for it. There will be future cases (ex - logs forwarded via fluentd/fluentbit) where the watcher won't directly forward data, but will need to publish information as to where the logs landed.

OK to keep this here for now.

File: pkg/watcher/reconciler/pipelinerun/controller.go [hunk 2]
```diff
@@ -28,17 +28,18 @@ import (
...
-func NewControllerWithConfig(ctx context.Context, client pb.ResultsClient, cfg *reconciler.Config) *controller.Impl {
+func NewControllerWithConfig(ctx context.Context, resultsClient pb.ResultsClient, cfg *reconciler.Config) *controller.Impl {
 	informer := pipelineruninformer.Get(ctx)
 	lister := informer.Lister()
 
 	c := &Reconciler{
 		LeaderAwareFuncs: leaderelection.NewLeaderAwareFuncs(lister.List),
-		client:           client,
+		resultsClient:    resultsClient,
+		logsClient:       logs.Get(ctx),
```
Comment (adambkaplan) [inline pkg/watcher/reconciler/pipelinerun/controller.go:42]: Non-blocking observation - some things we are wiring through the `Context` object, others we are not. Unsure if/how we can establish some consistency here.

File: cmd/api/main.go [hunk 3]
```diff
@@ -121,6 +122,9 @@ func main() {
 		grpc.UnaryInterceptor(prometheus.UnaryServerInterceptor),
 	)
 	v1alpha2pb.RegisterResultsServer(s, v1a2)
+	if configFile.LOGS_API {
+		v1alpha2pb.RegisterLogsServer(s, v1a2)
+	}
```
Comment (adambkaplan) [inline cmd/api/main.go:127]: This makes sense to me - if we want logs to be an opt-in feature, disable the API completely if it is not available.

**Review by adambkaplan — APPROVED** [review]
/approve

My only main comment is that in the future, we will need to create mechanisms for the watcher to store metadata of logs forwarded by other systems/services.

File: pkg/watcher/reconciler/pipelinerun/controller.go (hunk 2 above)
Comment (sayan-biswas) [reply pkg/watcher/reconciler/pipelinerun/controller.go:42]: All the other clients and informers used in Knative are injected in the controller context (as in other tekton controllers). I guess injected client is the consistent way when working with Knative.

File: cmd/api/README.md [hunk 4]
```diff
@@ -2,19 +2,21 @@
...
+| TLS_HOSTNAME_OVERRIDE | Override the hostname used to serve TLS. This should not be set (or set to the empty string) in production environments. | results.tekton.dev                           |
+| TLS_PATH              | Path to TLS files                                                                                                        | /etc/tls                                     |
 | LOG_LEVEL            | Log level for api server   | INFO                                         |
-| TLS_HOSTNAME_OVERRIDE| Override the hostname used to serve TLS. This should not be set (or set to the empty string) in production environments.     | results.tekton.dev                           |
-| TLS_PATH             | Path to TLS files          | /etc/tls                                     |
+| LOGS_API              | Enable logs storage service                                                                                              | false (default)                              |
+| Environment Variable | Description                | Example                                      |
```
Comment (khrm) [inline cmd/api/README.md:19]: Remove this line in the doc.

File: pkg/watcher/reconciler/pipelinerun/controller.go (hunk 2 above)
Comment (khrm) [reply pkg/watcher/reconciler/pipelinerun/controller.go:42]: Yes. We can try moving to knative/pkg but it's okay for now. We did this with the Trigger's EventListener sink HTTP server. The first implementation wasn't Knative. I will look into doing this with GRPC.

Comment (khrm) [reply pkg/watcher/reconciler/pipelinerun/controller.go:42]: Yes. We can try moving to knative/pkg but it's okay for now. We did this with the Trigger's EventListener sink HTTP server. The first implementation wasn't Knative. I will look into doing this with GRPC.

**Comment by adambkaplan** [comment]
/assign @dibyom 

/cc @alan-ghelardi 

Discussed with @sayan-biswas  - I closed #258, and instead of one big PR we think it's better to land it in pieces. This adds logs as an optional feature on the apiserver. If logs are disabled, the APIs are not served.

**Comment by alan-ghelardi** [comment]
Awesome, although I cannot approve yet.

**Review by dibyom — COMMENTED** [review]
/lgtm

**Review by khrm — COMMENTED** [review]
/kind feature
We need this label for release as discussed in the last WG call.

**Pull request merged**